src/compiler/build/build-stats.ts

```diff
@@ -0,0 +1,157 @@
+ * @param buildCtx
+ * @returns
+ */
+export function generateBuildStats(config: d.Config, buildCtx: d.BuildCtx) {
+  const buildResults = buildCtx.buildResults;
+  let jsonData: any;
+
+  if (buildResults.hasError) {
+    jsonData = {
+      diagnostics: buildResults.diagnostics,
+    };
+  } else {
+    try {
+      const stats: d.CompilerBuildStats = {
+        timestamp: buildResults.timestamp,
+        compiler: {
+          name: config.sys.name,
+          version: config.sys.version,
+        },
+        app: {
+          namespace: config.namespace,
+          fsNamespace: config.fsNamespace,
+          components: Object.keys(buildResults.componentGraph).length,
+          entries: Object.keys(buildResults.componentGraph).length,
+          bundles: buildResults.outputs.reduce((total, en) => (total += en.files.length), 0),
+          outputs: getAppOutputs(config, buildResults),
+        },
+        options: {
+          minifyJs: config.minifyJs,
+          minifyCss: config.minifyCss,
+          hashFileNames: config.hashFileNames,
+          hashedFileNameLength: config.hashedFileNameLength,
+          buildEs5: config.buildEs5,
+        },
+        esmBrowser: sanitizeBundlesForStats(buildCtx.esmBrowserComponentBundle),
+        esm: sanitizeBundlesForStats(buildCtx.esmComponentBundle),
+        es5: sanitizeBundlesForStats(buildCtx.es5ComponentBundle),
+        system: sanitizeBundlesForStats(buildCtx.systemComponentBundle),
+        commonjs: sanitizeBundlesForStats(buildCtx.commonJsComponentBundle),
+        components: getComponentsFileMap(config, buildCtx),
+        entries: buildCtx.entryModules,
+        sourceGraph: buildResults.componentGraph,
+        rollupResults: buildCtx.rollupResults,
+        collections: getCollections(config, buildCtx),
+      };
+
+      jsonData = stats;
+    } catch (e) {
+      console.log(e);
+    }
+  }
+
+  return jsonData;
+}
+
+/**
+ *
+ * @param config
+ * @param buildCtx
+ * @returns
+ */
+export async function writeBuildStats(config: d.Config, data: d.CompilerBuildStats) {
+  const statsTargets = config.outputTargets.filter((o) => o.type === 'stats') as d.OutputTargetStats[];
+
+  await Promise.all(
+    statsTargets.map((outputTarget) => {
+      config.sys.writeFile(outputTarget.file, JSON.stringify(data, null, 2));
+    })
+  );
+}
+
+function sanitizeBundlesForStats(bundleArray: d.BundleModule[]): d.CompilerBuildStatBundle[] {
+  if (!bundleArray) {
+    return [];
+  }
+
+  return bundleArray.map((bundle) => {
+    return {
+      // Should probably get the file size too.
+      key: bundle.entryKey,
+      components: bundle.cmps.map((c) => c.tagName),
+      bundleId: bundle.output.bundleId,
+      fileName: bundle.output.fileName,
+      imports: bundle.rollupResult.imports,
+      // code: bundle.rollupResult.code, // (use this to debug)
+      // Currently, this number is inaccurate vs what seems to be on disk.
+      byteSize: byteSize(bundle.rollupResult.code),
```

Comment:
@rwaskiewicz or @lars, any recs about how to get this number to be more accurate without reading the file from disk? Should I omit for the time being?

Reply:
How inaccurate are we talking? Like an order of magnitude? or a few bytes?

Reply:
Oh, duh. It's because the files are compressed. It's a symptom of uglifying the files being applied after the build. Ironically, there's also some larger discrepancy between esmBrowser and esm as a result of esm not being minified. Since esm is not compressed right now (there's another bug somewhere filed about this...) the esm file's byte sizes produced at this point are accurate. 

I think I'm going to rename this to "originalByteSize" instead. We should in the future figure out how to add that. Feels like an ordering problem. I'd hate to read files, as generating the stats may be _much_ slower.